File: labs/final-workspace/maps/random_map_gen.py
import numpy as np
import matplotlib.pyplot as plt
from heapq import heappush, heappop

def clear_area_around_point(map_array, point, radius, map_size):
    x_center, y_center = point
    for x in range(max(0, x_center - radius), min(map_size, x_center + radius + 1)):
        for y in range(max(0, y_center - radius), min(map_size, y_center + radius + 1)):
            if np.sqrt((x - x_center) ** 2 + (y - y_center) ** 2) <= radius:
                map_array[x, y] = 0  # Clear the area

def grow_obstacles(map_array, seed_points, max_growth_steps_range, growth_chance, map_size):
    max_growth_steps = np.random.randint(max_growth_steps_range[0], max_growth_steps_range[1] + 1)
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    for seed in seed_points:
        x, y = seed
        map_array[x, y] = 1  # Mark the seed as an obstacle
        for _ in range(np.random.randint(1, max_growth_steps + 1)):
            if np.random.rand() < growth_chance:
                dx, dy = directions[np.random.randint(0, len(directions))]
                new_x, new_y = x + dx, y + dy
                if 0 <= new_x < map_size and 0 <= new_y < map_size:
                    map_array[new_x, new_y] = 1
                    x, y = new_x, new_y  # Update the current position to new growth
    return map_array

def heuristic(a, b):
    return np.abs(a[0] - b[0]) + np.abs(a[1] - b[1])

def astar(map_array, start, end):
    neighbors = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    close_set = set()
    came_from = {}
    gscore = {start:0}
    fscore = {start:heuristic(start, end)}
    oheap = []
    heappush(oheap, (fscore[start], start))
    
    while oheap:
        current = heappop(oheap)[1]
        if current == end:
            data = []
            while current in came_from:
                data.append(current)
                current = came_from[current]
            return data[::-1]  # Return reversed path, from start to end
        close_set.add(current)
        for i, j in neighbors:
            neighbor = current[0] + i, current[1] + j            
            tentative_g_score = gscore[current] + heuristic(current, neighbor)
            if 0 <= neighbor[0] < map_array.shape[0] and 0 <= neighbor[1] < map_array.shape[1]:
                if map_array[neighbor[0]][neighbor[1]] == 1:
                    continue
                if neighbor in close_set and tentative_g_score >= gscore.get(neighbor, 0):
                    continue
                if tentative_g_score < gscore.get(neighbor, 0) or neighbor not in [i[1] for i in oheap]:
                    came_from[neighbor] = current
                    gscore[neighbor] = tentative_g_score
                    fscore[neighbor] = tentative_g_score + heuristic(neighbor, end)
                    heappush(oheap, (fscore[neighbor], neighbor))
    return False

def fill_enclosed_areas(map_array):
    map_size = map_array.shape[0]
    visited = np.zeros_like(map_array, dtype=bool)

    def flood_fill(x, y):
        stack = [(x, y)]
        while stack:
            x, y = stack.pop()
            if x < 0 or x >= map_size or y < 0 or y >= map_size:
                continue
            if visited[x, y] or map_array[x, y] == 1:
                continue
            visited[x, y] = True
            directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
            for dx, dy in directions:
                stack.append((x + dx, y + dy))

    # Start flood fill from all border cells that are empty
    for x in range(map_size):
        if not visited[x, 0] and map_array[x, 0] == 0:
            flood_fill(x, 0)
        if not visited[x, map_size - 1] and map_array[x, map_size - 1] == 0:
            flood_fill(x, map_size - 1)
    for y in range(map_size):
        if not visited[0, y] and map_array[0, y] == 0:
            flood_fill(0, y)
        if not visited[map_size - 1, y] and map_array[map_size - 1, y] == 0:
            flood_fill(map_size - 1, y)

    # Fill all unvisited, empty cells
    for x in range(map_size):
        for y in range(map_size):
            if map_array[x, y] == 0 and not visited[x, y]:
                map_array[x, y] = 1

def make_path_wide(map_array, path, width, map_size):
    radius = width // 2
    for point in path:
        clear_area_around_point(map_array, point, radius, map_size)

def calculate_distance(point1, point2):
    return np.sqrt((point2[0] - point1[0]) ** 2 + (point2[1] - point1[1]) ** 2)

def generate_map_with_random_shapes(size, num_seeds, max_growth_steps_range, growth_chance, path_width, attempts=25):
    if attempts <= 0:
        print("Failed to generate a valid map after maximum attempts.")
        return None, None, None, None
    
    map_array = np.zeros((size, size), dtype=np.int8)
    seed_points = [(np.random.randint(0, size), np.random.randint(0, size)) for _ in range(num_seeds)]
    
    # Pick start and end points ensure they are 3/4 of the map size away from each other
    min_distance = 0.75 * size
    while True:
        start_point = (np.random.randint(8, size-8), np.random.randint(8, size-8))
        end_point = (np.random.randint(8, size-8), np.random.randint(8, size-8))
        
        distance = calculate_distance(start_point, end_point)
        
        if distance > min_distance:
            break
    
    # Grow obstacles from seeds
    map_array = grow_obstacles(map_array, seed_points, max_growth_steps_range, growth_chance, size)

    # Fill enclosed areas
    fill_enclosed_areas(map_array)
    
    # Clear areas around start and end points before growing obstacles
    clear_area_around_point(map_array, start_point, 8, size)
    clear_area_around_point(map_array, end_point, 8, size)

    # Ensure the start and end points are not obstacles
    map_array[start_point] = 0
    map_array[end_point] = 0
    
    # Find a path between start and end points
    path = astar(map_array, start_point, end_point)
    if path:
        make_path_wide(map_array, path, path_width, size)  # Make the path wide
        return map_array, start_point, end_point, path
    else:
        print(f"No valid path found, retrying... ({25 - attempts + 1}/25)")
        return generate_map_with_random_shapes(size, num_seeds, max_growth_steps_range, growth_chance, path_width, attempts - 1)

def invert_map(map_array):
    # Invert the map: obstacles become free spaces (1 becomes 0, 0 becomes 1)
    inverted_map = 1 - map_array
    return inverted_map

# Configuration
map_size = 128
num_seeds = 25  # Number of obstacle seeds
max_growth_steps_range = (1500, 2000)  # Range of growth steps for obstacles
growth_chance = 0.7
path_width = 10

map_array_with_random_shapes, start_point, end_point, path = generate_map_with_random_shapes(map_size, num_seeds, max_growth_steps_range, growth_chance, path_width)


# map_array_with_random_shapes = invert_map(map_array_with_random_shapes)

if map_array_with_random_shapes is not None:
    plt.figure(figsize=(10, 10))
    plt.imshow(map_array_with_random_shapes, cmap='binary')
     # Extract X and Y coordinates from the path
    x_coords = [point[0] for point in path]
    y_coords = [point[1] for point in path]

    # Plot the path
    plt.plot(y_coords, x_coords, color="red", linewidth=2)  # Plot path in yellow for visibility
    plt.scatter([start_point[1]], [start_point[0]], color='green')  # Start point
    plt.scatter([end_point[1]], [end_point[0]], color='blue')  # End point
    plt.title("Map with A* Path from Start to End")
    plt.show()
else:
    print("Failed to find a path.")

File: labs/final-workspace/maps/randomMapGen.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <string.h>
#include <limits.h>
#include <stdbool.h>

// Structs
typedef struct
{
    int x, y;
} Point;

typedef struct
{
    Point point;
    int gScore;
    int fScore;
    Point cameFrom;
} Node;

int find_index_in_open_set(Node openSet[], int openSetSize, Point point) {
    for (int i = 0; i < openSetSize; ++i) {
        if (openSet[i].point.x == point.x && openSet[i].point.y == point.y) {
            return i;
        }
    }
    return -1; // Not found
}

void add_to_open_set(Node openSet[], int *openSetSize, Node node) {
    openSet[(*openSetSize)++] = node;
}

int compare_nodes(const void *a, const void *b) {
    Node *nodeA = (Node *)a;
    Node *nodeB = (Node *)b;
    return nodeA->fScore - nodeB->fScore;
}

void sort_open_set(Node openSet[], int openSetSize) {
    qsort(openSet, openSetSize, sizeof(Node), compare_nodes);
}

// // Function to add a node to the priority queue (simplified insertion sort for demonstration purposes)
// void push(Node *heap, int *heapSize, Node node)
// {
//     heap[(*heapSize)++] = node;
//     // Simple insertion sort to keep the queue sorted by fScore
//     for (int i = *heapSize - 1; i > 0 && heap[i].fScore < heap[i - 1].fScore; i--)
//     {
//         Node temp = heap[i];
//         heap[i] = heap[i - 1];
//         heap[i - 1] = temp;
//     }
// }

// // Function to pop the node with the lowest fScore from the priority queue
// Node pop(Node *heap, int *heapSize)
// {
//     Node node = heap[0];
//     memmove(&heap[0], &heap[1], (--(*heapSize)) * sizeof(Node)); // Shift elements down
//     return node;
// }

// int inSet(Point *set, int setSize, Point point)
// {
//     for (int i = 0; i < setSize; i++)
//     {
//         if (set[i].x == point.x && set[i].y == point.y)
//             return 1;
//     }
//     return 0;
// }

// // Check if a point is in the set
// int inNodeSet(Node *set, int setSize, Point point)
// {
//     for (int i = 0; i < setSize; i++)
//     {
//         if (set[i].point.x == point.x && set[i].point.y == point.y)
//             return 1;
//     }
//     return 0;
// }

// Function to initialize the map array
void init_map(int **map, int size)
{
    for (int i = 0; i < size; i++)
    {
        for (int j = 0; j < size; j++)
        {
            map[i][j] = 0; // Initialize all cells to 0
        }
    }
}

// Function to print the map
void print_map(int **map, int size)
{
    for (int i = 0; i < size; i++)
    {
        for (int j = 0; j < size; j++)
        {
            printf("%d ", map[i][j]);
        }
        printf("\n");
    }
}

// Initialize random seed
void init_random()
{
    srand((unsigned)time(NULL));
}

// Generate a random integer between min and max (inclusive)
int random_int(int min, int max)
{
    return min + rand() % (max - min + 1);
}

// Generate a random float between 0.0 and 1.0
float random_float()
{
    return (float)rand() / (float)RAND_MAX;
}

//  Heuristic for A* algorithm
int heuristic(Point a, Point b)
{
    return abs(a.x - b.x) + abs(a.y - b.y);
}

// // Function to mark the path on the map for visualization
// void mark_path(int **map, Point *path, int pathSize)
// {
//     for (int i = 0; i < pathSize; i++)
//     {
//         map[path[i].x][path[i].y] = 2; // Marking the path with '2'
//     }
// }


void print_map_with_path(int **map, Point path[], int pathSize, int map_size) {
    printf("Map with path:\n");
    for (int i = 0; i < map_size; ++i) {
        for (int j = 0; j < map_size; ++j) {
            char cell = '.';
            for (int k = 0; k < pathSize; ++k) {
                if (path[k].x == i && path[k].y == j) {
                    cell = 'P'; // Mark path
                    break;
                }
            }
            if (map[i][j] == 1) cell = '#'; // Mark obstacle
            printf("%c ", cell);
        }
        printf("\n");
    }
}

// Functions
void clear_area_around_point(int **map_array, int x_center, int y_center, int radius, int map_size)
{
    for (int x = fmax(0, x_center - radius); x <= fmin(map_size - 1, x_center + radius); ++x)
    {
        for (int y = fmax(0, y_center - radius); y <= fmin(map_size - 1, y_center + radius); ++y)
        {
            if (sqrt((x - x_center) * (x - x_center) + (y - y_center) * (y - y_center)) <= radius)
            {
                map_array[x][y] = 0; // Clear the area
            }
        }
    }
}
void grow_obstacles(int **map_array, Point *seed_points, int num_seeds, int min_growth_steps, int max_growth_steps, float growth_chance, int map_size)
{
    int max_growth_steps_range = random_int(min_growth_steps, max_growth_steps);
    Point directions[4] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    for (int i = 0; i < num_seeds; i++)
    {
        int x = seed_points[i].x;
        int y = seed_points[i].y;
        map_array[x][y] = 1; // Mark the seed as an obstacle
        int steps = random_int(1, max_growth_steps_range);
        for (int step = 0; step < steps; step++)
        {
            if (random_float() < growth_chance)
            {
                int dir_index = random_int(0, 3);
                int dx = directions[dir_index].x;
                int dy = directions[dir_index].y;
                int new_x = x + dx;
                int new_y = y + dy;
                if (0 <= new_x && new_x < map_size && 0 <= new_y && new_y < map_size)
                {
                    map_array[new_x][new_y] = 1;
                    x = new_x;
                    y = new_y; // Update the current position to new growth
                }
            }
        }
    }
}
int astar(int **map_array, Point start, Point end, Point *path, int map_size)
{
    Node openSet[map_size * map_size]; // Open set as simple array
    int openSetSize = 0;

    Node nodes[map_size][map_size]; // Node information for each point

    for (int x = 0; x < map_size; ++x) {
        for (int y = 0; y < map_size; ++y) {
            nodes[x][y].point.x = x;
            nodes[x][y].point.y = y;
            nodes[x][y].gScore = INT_MAX;
            nodes[x][y].fScore = INT_MAX;
            nodes[x][y].cameFrom.x = -1; // -1 indicates 'undefined'
            nodes[x][y].cameFrom.y = -1;
        }
    }

    // Initialize start node
    nodes[start.x][start.y].gScore = 0;
    nodes[start.x][start.y].fScore = heuristic(start, end);
    add_to_open_set(openSet, &openSetSize, nodes[start.x][start.y]);

    Point directions[4] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    while (openSetSize > 0) {
        Node current = openSet[0]; // Node with the lowest fScore
        if (current.point.x == end.x && current.point.y == end.y) {
            // Reconstruct path
            int pathSize = 0;
            while (!(current.point.x == start.x && current.point.y == start.y)) {
                path[pathSize++] = current.point;
                current = nodes[current.cameFrom.x][current.cameFrom.y];
            }
            path[pathSize++] = start; // Include start in the path
            // Reverse path
            for (int i = 0; i < pathSize / 2; ++i) {
                Point temp = path[i];
                path[i] = path[pathSize - 1 - i];
                path[pathSize - 1 - i] = temp;
            }
            return pathSize;
        }

        // Move current Node from Open Set to Closed Set
        memmove(openSet, openSet + 1, (--openSetSize) * sizeof(Node)); // Remove current
        sort_open_set(openSet, openSetSize); // Sort again after removal, simplistic approach

        for (int i = 0; i < 4; ++i) { // Check all four neighbors
            Point nextPoint = {current.point.x + directions[i].x, current.point.y + directions[i].y};
            if (nextPoint.x < 0 || nextPoint.x >= map_size || nextPoint.y < 0 || nextPoint.y >= map_size) continue; // Skip if out of bounds
            if (map_array[nextPoint.x][nextPoint.y] == 1) continue; // Skip obstacles

            int tentative_gScore = current.gScore + 1; // Assume cost between any two nodes is 1
            if (tentative_gScore < nodes[nextPoint.x][nextPoint.y].gScore) {
                // This path is better than any previous one. Record it!
                nodes[nextPoint.x][nextPoint.y].cameFrom = current.point;
                nodes[nextPoint.x][nextPoint.y].gScore = tentative_gScore;
                nodes[nextPoint.x][nextPoint.y].fScore = tentative_gScore + heuristic(nextPoint, end);

                if (find_index_in_open_set(openSet, openSetSize, nextPoint) == -1) {
                    add_to_open_set(openSet, &openSetSize, nodes[nextPoint.x][nextPoint.y]); // Add to open set if not already present
                    sort_open_set(openSet, openSetSize); // Keep open set sorted
                }
            }
        }
    }

    return 0; // Path not found
}

// Test functions
void clear_area_around_point_test()
{
    int map_size = 10;              // Example map size
    int radius = 3;                 // Example radius
    int x_center = 5, y_center = 5; // Example center point

    // Allocate memory for the map array
    int **map_array = (int **)malloc(map_size * sizeof(int *));
    for (int i = 0; i < map_size; i++)
    {
        map_array[i] = (int *)malloc(map_size * sizeof(int));
        for (int j = 0; j < map_size; j++)
        {
            map_array[i][j] = 1; // Initialize map with non-zero values to see the clearing effect
        }
    }

    // Clear area around point
    clear_area_around_point(map_array, x_center, y_center, radius, map_size);

    // Print the map array to verify the clearing
    for (int i = 0; i < map_size; i++)
    {
        for (int j = 0; j < map_size; j++)
        {
            printf("%d ", map_array[i][j]);
        }
        printf("\n");
    }

    // Free the allocated memory
    for (int i = 0; i < map_size; i++)
    {
        free(map_array[i]);
    }
    free(map_array);
    return;
}
void grow_obstacles_test()
{
    int MAP_SIZE = 20;
    init_random(); // Initialize the random number generator

    // Allocate memory for the map
    int **map = (int **)malloc(MAP_SIZE * sizeof(int *));
    for (int i = 0; i < MAP_SIZE; i++)
    {
        map[i] = (int *)malloc(MAP_SIZE * sizeof(int));
    }

    init_map(map, MAP_SIZE); // Initialize map

    // Define seed points
    Point seed_points[] = {{5, 5}, {10, 10}, {15, 15}};
    int num_seeds = sizeof(seed_points) / sizeof(seed_points[0]);

    // Call the grow_obstacles function
    grow_obstacles(map, seed_points, num_seeds, 50, 75, 0.75, MAP_SIZE);

    // Print the map to see the obstacles
    print_map(map, MAP_SIZE);

    // Free allocated memory
    for (int i = 0; i < MAP_SIZE; i++)
    {
        free(map[i]);
    }
    free(map);

    return;
}
void astar_test()
{
    int MAP_SIZE = 20;
    init_random(); // Initialize the random number generator

    // Allocate memory for the map
    int **map = (int **)malloc(MAP_SIZE * sizeof(int *));
    for (int i = 0; i < MAP_SIZE; i++)
    {
        map[i] = (int *)malloc(MAP_SIZE * sizeof(int));
    }

    init_map(map, MAP_SIZE); // Initialize map

    // Example: setting up a simple obstacle
    for (int i = 2; i < 8; ++i) {
        map[i][5] = 1; // Vertical wall, except at one gap
    }

    Point start = {0, 0}; // Starting point
    Point end = {9, 9}; // Ending point or goal
    Point path[MAP_SIZE * MAP_SIZE]; // Allocate space for the path
    int pathSize;

    pathSize = astar(map, start, end, path, MAP_SIZE);

    if (pathSize > 0) {
        printf("Path found, size: %d\n", pathSize);
        print_map_with_path(map, path, pathSize, MAP_SIZE);
    } else {
        printf("No path found.\n");
    }
}

int main()
{
    astar_test();
    return 0;
}